lib/storage/storage-local.js

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2018, Compiler Explorer Authors
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+const StorageBase = require('./storage').StorageBase,
+    logger = require('../logger').logger,
+    fs = require('fs-extra'),
+    _ = require('underscore'),
+    path = require('path');
+
+const MIN_STORED_ID_LENGTH = 6;
+
+class StorageLocal extends StorageBase {
+    constructor(compilerProps) {
+        super(compilerProps);
+        this.storageFolder = path.normalize(compilerProps.ceProps('localStorageFolder', './lib/storage/data/'));
+        // Ensure we have a working storage dir before we have a chance to process anything
+        fs.ensureDirSync(this.storageFolder);
+        logger.info(`Using local storage solution on ${this.storageFolder}`);
+    }
+
+    storeItem(item) {
+        const filePath = path.join(this.storageFolder, item.uniqueSubHash);
+        return fs.writeJson(filePath, item, {encoding: 'utf8'});
+    }
+
+    findUniqueSubhash(hash) {
+        logger.info(`Finding local unique subhash for ${hash}`);
+        // This currently works on a hardcoded, local directory.
+        return fs.readdir(this.storageFolder)
```

Comment:
Could you explain what this code is doing? Im trying to make heads or tail of it and not having much luck. 

When I see hash and a forloop I always wonder if we could improve :)

Reply:
This tries to find the shortest unique subhash for which there is nothing stored yet with that name, or if something is stored, their full hashes match. (New object being stored/Storing already stored object)
This is to make URLS ~6 characters long instead of ~85

Reply:
@jaredwy channelling Sean Parent here? :) I was mulling over this and I'm not sure how it could be done more functionally. If you have a few moments to suggest change that'd be great. I'd learn here.

Reply:
Ah just couldn't make sense of the name and then having a loop. 
Now that I know what hte code is doing, I can see if I can find a way to make it a bit easier to understand :)